22/2.rs

```diff
@@ -0,0 +1,267 @@
+type Solution = usize;
+
+const EXAMPLES: [(&str, Solution); 3] = [
+    ("1", 39),
+    ("2", 39769202357779),
+    ("3", 2758514936282235),
+];
+
+const DAY: u8 = 22;
+
+fn main() {
+    let results = EXAMPLES.iter()
+        .zip(
+            EXAMPLES.iter()
+                .map(|t| format!("{}/{}.ie", DAY, t.0))
+                .map(operation)
+        )
+        .map(|((name, expected), result)| 
+            (
+                *name,
+                result
+                    .and_then(|actual| if *expected == actual {
+                        return Ok(());
+                    } else {
+                        return Err(format!("Expected {} but got {}", expected, actual));
+                    })
+            )
+        )
+        .collect::<Vec<(&str, Result<(), String>)>>();
+    results.iter()
+        .for_each(|(name, result)| match result {
+            Ok(()) => println!("Example {} passed.", name),
+            Err(msg) => println!("Example {} failed: {}.", name, msg)
+        });
+
+    if results.iter().any(|t| t.1.is_err()) {
+        panic!("Please address errors before attempting the problem.")
+    }
+
+    println!(
+        "{}",
+        operation(format!("{}/input", DAY)).expect("Unexpected Error in main input.")
+    );
+}
+
+// fn error<T>(msg: &str) -> Result<T, String> {
+//     return Err(String::from(msg));
+// }
+
+fn operation(filename: String) -> Result<Solution, String> {
+    return fs::read_to_string(filename)
+        .map_err(|io_error| format!("{}", io_error))
+        .and_then(parse)
+        .and_then(solve);
+}
+
+#[derive(PartialEq, Eq)]
+struct Command {
+    on: bool,
+    points: Points
+}
+
+impl Command {
+    fn from_input(line: &str) -> Result<Command, String> {
+        let mut bits = line.split(&[' ', ','][..]);
+        let on = bits.next()
+            .ok_or(String::from("No on/off"))
+            .and_then(|on| match on {
+                "on" => Ok(true),
+                "off" => Ok(false),
+                other => Err(format!("Unrecognized on/off: '{}'", other))
+            })?;
+        let x = bits.next()
+            .ok_or(String::from("insufficient segments for x range."))
+            .and_then(|s| parse_range("x=", s))?;
+        let y = bits.next()
+            .ok_or(String::from("insufficient segments for y range."))
+            .and_then(|s| parse_range("y=", s))?;
+        let z = bits.next()
+            .ok_or(String::from("insufficient segments for z range."))
+            .and_then(|s| parse_range("z=", s))?;
+        return Ok(Command {
+            on: on,
+            points: Points {
+                xs: x,
+                ys: y,
+                zs: z
+            }
+        })
+    }
+}
+
+fn parse_range(prefix: &str, line: &str) -> Result<Range, String> {
+    let bounds = line.strip_prefix(prefix)
+        .ok_or(format!("Expected prefix '{}' on '{}'", prefix, line))
+        .and_then(|rest| rest.split("..")
+            .map(str::parse)
```

Comment:
My AoC jam here is to do `.filter_map(parse().ok()))`, which converts the `Result` to an `Option`, then filters and unwraps everything. I do this mostly during AoC; it does throw away error information, so it's a "I'm real confident about my inputs" kind of move.

Reply:
Hah!  Yeah, I see that. I'm less unconfident in my inputs and more unconfident that I've strung the right rust together.